File: backend/src/routes/analytics.js
const router = require('express').Router();
const axios = require('axios');
const Product = require('../models/Product');
const Sale = require('../models/Sale');
const { buildFeatureVector } = require('../services/featureBuilder');

const ML_URL = process.env.ML_SERVICE_URL || 'http://localhost:5001';

async function mlGet(path, params = {}, timeout = 12000) {
  const res = await axios.get(`${ML_URL}${path}`, { params, timeout });
  return res.data;
}
async function mlPost(path, body = {}, timeout = 12000) {
  const res = await axios.post(`${ML_URL}${path}`, body, { timeout });
  return res.data;
}

// ── Local fallbacks ──────────────────────────────────────────────

async function localDecompose(productId, category, days = 90) {
  const since = new Date(Date.now() - days * 86400000);
  const sales = await Sale.find({ productId, timestamp: { $gte: since } })
    .select('quantity timestamp').lean();

  const dailyMap = {};
  for (const s of sales) {
    const d = new Date(s.timestamp).toISOString().slice(0, 10);
    dailyMap[d] = (dailyMap[d] || 0) + s.quantity;
  }

  const dates = [];
  const observed = [];
  for (let i = days - 1; i >= 0; i--) {
    const d = new Date(Date.now() - i * 86400000).toISOString().slice(0, 10);
    dates.push(d);
    observed.push(dailyMap[d] || 0);
  }

  const window = 7;
  const trend = observed.map((_, i) => {
    const slice = observed.slice(Math.max(0, i - window), i + window + 1);
    return slice.reduce((a, b) => a + b, 0) / slice.length;
  });
  const seasonal = observed.map((v, i) => v - trend[i]);
  const residual = observed.map((v, i) => v - trend[i] - seasonal[i]);

  return {
    dates, observed, trend: trend.map(v => +v.toFixed(2)),
    seasonal: seasonal.map(v => +v.toFixed(2)),
    residual: residual.map(v => +v.toFixed(2)),
    trend_direction: trend[trend.length - 1] > trend[0] ? 'increasing' : 'decreasing',
    seasonality_strength: 0.4,
    data_points: sales.length
  };
}

async function localAnomalies(productId, days = 60) {
  const since = new Date(Date.now() - days * 86400000);
  const sales = await Sale.find({ productId, timestamp: { $gte: since } })
    .select('quantity timestamp').lean();

  if (!sales.length) return { anomalies: [], anomaly_rate: 0, total_days: 0, flagged_days: 0, all_scores: [] };

  const dailyMap = {};
  for (const s of sales) {
    const d = new Date(s.timestamp).toISOString().slice(0, 10);
    dailyMap[d] = (dailyMap[d] || 0) + s.quantity;
  }

  const entries = Object.entries(dailyMap).map(([date, qty]) => ({ date, quantity: qty }));
  const qtys = entries.map(e => e.quantity);
  const mean = qtys.reduce((a, b) => a + b, 0) / qtys.length;
  const std = Math.sqrt(qtys.reduce((a, b) => a + (b - mean) ** 2, 0) / qtys.length);

  const anomalies = [];
  const all_scores = entries.map(e => {
    const zscore = std > 0 ? Math.abs(e.quantity - mean) / std : 0;
    const is_anomaly = zscore > 2.5;
    if (is_anomaly) anomalies.push({ date: e.date, quantity: e.quantity, anomaly_score: +(-zscore).toFixed(3), severity: zscore > 3.5 ? 'high' : 'medium' });
    return { date: e.date, score: +(-zscore / 5).toFixed(3), is_anomaly };
  });

  return { anomalies, anomaly_rate: +(anomalies.length / entries.length).toFixed(3), total_days: entries.length, flagged_days: anomalies.length, all_scores };
}

async function localElasticity(productId) {
  const since = new Date(Date.now() - 90 * 86400000);
  const sales = await Sale.find({ productId, timestamp: { $gte: since } })
    .select('quantity price').lean();

  if (sales.length < 10) return { elasticity: -1.2, interpretation: 'Estimated (insufficient data)', confidence: 'low', r2: 0, data_points: sales.length };

  const logP = sales.map(s => Math.log(Math.max(s.price, 0.01)));
  const logQ = sales.map(s => Math.log(Math.max(s.quantity, 0.1)));
  const n = logP.length;
  const meanP = logP.reduce((a, b) => a + b) / n;
  const meanQ = logQ.reduce((a, b) => a + b) / n;
  const num = logP.reduce((s, p, i) => s + (p - meanP) * (logQ[i] - meanQ), 0);
  const den = logP.reduce((s, p) => s + (p - meanP) ** 2, 0);
  const elasticity = den > 0 ? num / den : -1.2;

  const pred = logP.map(p => meanQ + elasticity * (p - meanP));
  const ss_res = logQ.reduce((s, q, i) => s + (q - pred[i]) ** 2, 0);
  const ss_tot = logQ.reduce((s, q) => s + (q - meanQ) ** 2, 0);
  const r2 = ss_tot > 0 ? 1 - ss_res / ss_tot : 0;

  const e = +elasticity.toFixed(3);
  const interpretation = e < -2 ? 'Highly elastic — demand very sensitive to price' :
    e < -1 ? 'Elastic — demand moderately sensitive to price' :
    e < 0 ? 'Inelastic — demand relatively insensitive to price' : 'Unusual — demand increases with price';

  return { elasticity: e, interpretation, r2: +r2.toFixed(3), confidence: r2 > 0.5 ? 'high' : r2 > 0.2 ? 'medium' : 'low', data_points: n };
}

async function localWhatif(productId, baseFeatures, scenarios) {
  const baseDemand = (baseFeatures.avg_daily_sales_30d || 1) * 30;
  const basePrice = baseFeatures.price || 50;
  const results = scenarios.map(s => {
    const newPrice = basePrice * (1 + (s.price_change || 0));
    const elasticity = -1.2;
    const demandMult = (newPrice / basePrice) ** elasticity;
    const trendMult = 1 + ((s.trend_change || 0) / 200);
    const newDemand = baseDemand * demandMult * trendMult;
    const demandChangePct = (newDemand - baseDemand) / baseDemand * 100;
    const newRevenue = newDemand * newPrice;
    const revenueChangePct = (newRevenue - baseDemand * basePrice) / (baseDemand * basePrice) * 100;
    return {
      scenario_name: s.name, price_change_pct: +((s.price_change || 0) * 100).toFixed(1),
      new_price: +newPrice.toFixed(2), predicted_demand: +newDemand.toFixed(1),
      demand_change_pct: +demandChangePct.toFixed(1), new_revenue: +newRevenue.toFixed(2),
      revenue_change_pct: +revenueChangePct.toFixed(1),
      recommendation: revenueChangePct > 0 ? 'Proceed' : 'Caution'
    };
  });
  return { base_demand: +baseDemand.toFixed(1), base_price: basePrice, base_revenue: +(baseDemand * basePrice).toFixed(2), scenarios: results };
}

// ── Routes ───────────────────────────────────────────────────────

router.get('/decompose/:productId', async (req, res) => {
  try {
    const product = await Product.findById(req.params.productId);
    if (!product) return res.status(404).json({ success: false, error: 'Product not found' });
    try {
      const data = await mlGet(`/analytics/decompose/${req.params.productId}`, { category: product.category, days: req.query.days || 90 });
      return res.json({ success: true, data });
    } catch (_) {
      const data = await localDecompose(req.params.productId, product.category, parseInt(req.query.days) || 90);
      return res.json({ success: true, data, fallback: true });
    }
  } catch (err) { res.status(500).json({ success: false, error: err.message }); }
});

router.get('/anomalies/:productId', async (req, res) => {
  try {
    try {
      const data = await mlGet(`/analytics/anomalies/${req.params.productId}`, { days: req.query.days || 60 });
      return res.json({ success: true, data });
    } catch (_) {
      const data = await localAnomalies(req.params.productId, parseInt(req.query.days) || 60);
      return res.json({ success: true, data, fallback: true });
    }
  } catch (err) { res.status(500).json({ success: false, error: err.message }); }
});

router.get('/elasticity/:productId', async (req, res) => {
  try {
    try {
      const data = await mlGet(`/analytics/elasticity/${req.params.productId}`);
      return res.json({ success: true, data });
    } catch (_) {
      const data = await localElasticity(req.params.productId);
      return res.json({ success: true, data, fallback: true });
    }
  } catch (err) { res.status(500).json({ success: false, error: err.message }); }
});

router.post('/whatif', async (req, res) => {
  try {
    const { productId, scenarios } = req.body;
    const product = await Product.findById(productId);
    if (!product) return res.status(404).json({ success: false, error: 'Product not found' });
    const features = await buildFeatureVector(product, null, product.price);
    try {
      const data = await mlPost('/analytics/whatif', { product_id: productId, category: product.category, base_features: features, scenarios });
      return res.json({ success: true, data });
    } catch (_) {
      const data = await localWhatif(productId, features, scenarios);
      return res.json({ success: true, data, fallback: true });
    }
  } catch (err) { res.status(500).json({ success: false, error: err.message }); }
});

router.get('/monitor/performance', async (req, res) => {
  try {
    const data = await mlGet('/monitor/performance', { days: req.query.days || 30 });
    res.json({ success: true, data });
  } catch (_) {
    res.json({ success: true, data: { records: 0, message: 'No feedback data yet.' } });
  }
});

router.get('/monitor/drift', async (req, res) => {
  try {
    const data = await mlGet('/monitor/drift');
    res.json({ success: true, data });
  } catch (_) {
    res.json({ success: true, data: { drift_detected: false, psi: 0, alerts: ['Insufficient data for drift detection'], recommendation: 'Collect more predictions first' } });
  }
});

router.get('/monitor/audit', async (req, res) => {
  try {
    const data = await mlGet('/monitor/audit', { product_id: req.query.productId, limit: req.query.limit || 50 });
    res.json({ success: true, data: data.data });
  } catch (_) {
    res.json({ success: true, data: [] });
  }
});

router.post('/monitor/feedback', async (req, res) => {
  try {
    const data = await mlPost('/monitor/feedback', req.body);
    res.json({ success: true, data });
  } catch (err) { res.status(500).json({ success: false, error: err.message }); }
});

router.post('/train/ppo', async (req, res) => {
  try {
    const data = await mlPost('/train/ppo', {}, 300000);
    res.json({ success: true, data });
  } catch (err) { res.status(500).json({ success: false, error: err.message }); }
});

module.exports = router;

File: ml-service/model/analytics.py
"""
Advanced analytics:
- STL decomposition (trend + seasonality + residual)
- Isolation Forest anomaly detection
- Price elasticity estimation
- Cross-product correlation
- What-if price simulator
"""
import os
import numpy as np
import pandas as pd
from datetime import datetime, timedelta

try:
    from statsmodels.tsa.seasonal import STL
    STATSMODELS_AVAILABLE = True
except ImportError:
    STATSMODELS_AVAILABLE = False

try:
    from sklearn.ensemble import IsolationForest
    SKLEARN_AVAILABLE = True
except ImportError:
    SKLEARN_AVAILABLE = False

from pymongo import MongoClient
from dotenv import load_dotenv

load_dotenv()
MONGO_URI = os.getenv("MONGO_URI")


def _get_sales_series(product_id: str, days: int = 90):
    """Fetch daily sales series for a product."""
    try:
        from bson import ObjectId
        client = MongoClient(MONGO_URI, serverSelectionTimeoutMS=5000)
        db = client["demandforecast"]
        since = datetime.utcnow() - timedelta(days=days)
        sales = list(db.sales.find(
            {"productId": ObjectId(product_id), "timestamp": {"$gte": since}},
            {"quantity": 1, "timestamp": 1, "price": 1}
        ))
        client.close()
        return sales
    except Exception:
        return []


def stl_decompose(product_id: str, category: str, days: int = 90) -> dict:
    """
    STL decomposition of demand time series.
    Returns trend, seasonal, and residual components.
    """
    sales = _get_sales_series(product_id, days)

    if len(sales) < 14:
        return _synthetic_decomposition(category, days)

    # Build daily series
    df = pd.DataFrame(sales)
    df["date"] = pd.to_datetime(df["timestamp"]).dt.date
    daily = df.groupby("date")["quantity"].sum().reset_index()
    daily = daily.set_index("date")

    # Fill missing days with 0
    idx = pd.date_range(daily.index.min(), daily.index.max(), freq="D")
    daily = daily.reindex(idx, fill_value=0)
    series = daily["quantity"].values.astype(float)

    if not STATSMODELS_AVAILABLE or len(series) < 14:
        return _manual_decomposition(series, daily.index)

    try:
        stl = STL(series, period=7, robust=True)
        result = stl.fit()
        dates = [str(d.date()) for d in daily.index]
        return {
            "dates": dates,
            "observed": [round(float(v), 2) for v in series],
            "trend": [round(float(v), 2) for v in result.trend],
            "seasonal": [round(float(v), 2) for v in result.seasonal],
            "residual": [round(float(v), 2) for v in result.resid],
            "trend_direction": "increasing" if result.trend[-1] > result.trend[0] else "decreasing",
            "seasonality_strength": round(float(np.std(result.seasonal) / (np.std(result.seasonal) + np.std(result.resid) + 1e-8)), 3),
            "data_points": len(series)
        }
    except Exception as e:
        return _manual_decomposition(series, daily.index)


def _manual_decomposition(series, index):
    """Simple moving average decomposition fallback."""
    n = len(series)
    window = min(7, n // 2)
    trend = pd.Series(series).rolling(window, center=True, min_periods=1).mean().values
    seasonal = series - trend
    residual = series - trend - seasonal
    dates = [str(d.date()) if hasattr(d, 'date') else str(d) for d in index]
    return {
        "dates": dates,
        "observed": [round(float(v), 2) for v in series],
        "trend": [round(float(v), 2) for v in trend],
        "seasonal": [round(float(v), 2) for v in seasonal],
        "residual": [round(float(v), 2) for v in residual],
        "trend_direction": "increasing" if trend[-1] > trend[0] else "decreasing",
        "seasonality_strength": round(float(np.std(seasonal) / (np.std(seasonal) + np.std(residual) + 1e-8)), 3),
        "data_points": n
    }


def _synthetic_decomposition(category: str, days: int):
    """Generate synthetic decomposition for cold start."""
    base = {"Electronics": 3, "Clothing": 6, "Food": 15, "Furniture": 1, "Books": 4, "Toys": 4}.get(category, 3)
    dates = [(datetime.utcnow() - timedelta(days=days-i)).strftime("%Y-%m-%d") for i in range(days)]
    trend = [base * (1 + i * 0.002) for i in range(days)]
    seasonal = [base * 0.3 * np.sin(i * 2 * np.pi / 7) for i in range(days)]
    observed = [max(0, t + s + np.random.normal(0, base * 0.1)) for t, s in zip(trend, seasonal)]
    residual = [o - t - s for o, t, s in zip(observed, trend, seasonal)]
    return {
        "dates": dates, "observed": [round(v, 2) for v in observed],
        "trend": [round(v, 2) for v in trend], "seasonal": [round(v, 2) for v in seasonal],
        "residual": [round(v, 2) for v in residual],
        "trend_direction": "increasing", "seasonality_strength": 0.45, "data_points": 0
    }


def detect_anomalies(product_id: str, days: int = 60) -> dict:
    """
    Isolation Forest anomaly detection on sales data.
    Returns anomaly scores and flagged dates.
    """
    sales = _get_sales_series(product_id, days)

    if len(sales) < 10:
        return {"anomalies": [], "anomaly_rate": 0, "message": "Insufficient data"}

    df = pd.DataFrame(sales)
    df["date"] = pd.to_datetime(df["timestamp"]).dt.date
    df["hour"] = pd.to_datetime(df["timestamp"]).dt.hour
    daily = df.groupby("date").agg(
        quantity=("quantity", "sum"),
        transactions=("quantity", "count"),
        avg_price=("price", "mean")
    ).reset_index()

    if len(daily) < 5:
        return {"anomalies": [], "anomaly_rate": 0, "message": "Insufficient daily data"}

    features = daily[["quantity", "transactions"]].values

    if not SKLEARN_AVAILABLE:
        return {"anomalies": [], "anomaly_rate": 0, "message": "sklearn not available"}

    iso = IsolationForest(contamination=0.1, random_state=42, n_estimators=100)
    labels = iso.fit_predict(features)
    scores = iso.score_samples(features)

    anomalies = []
    for i, (label, score) in enumerate(zip(labels, scores)):
        if label == -1:
            anomalies.append({
                "date": str(daily.iloc[i]["date"]),
                "quantity": int(daily.iloc[i]["quantity"]),
                "anomaly_score": round(float(score), 3),
                "severity": "high" if score < -0.3 else "medium"
            })

    return {
        "anomalies": anomalies,
        "anomaly_rate": round(len(anomalies) / len(daily), 3),
        "total_days": len(daily),
        "flagged_days": len(anomalies),
        "all_scores": [{"date": str(daily.iloc[i]["date"]), "score": round(float(scores[i]), 3), "is_anomaly": labels[i] == -1} for i in range(len(daily))]
    }


def price_elasticity(product_id: str) -> dict:
    """
    Estimate price elasticity of demand using log-log regression.
    Elasticity = % change in demand / % change in price
    """
    sales = _get_sales_series(product_id, 90)

    if len(sales) < 20:
        return {"elasticity": -1.2, "interpretation": "Estimated (insufficient data)", "confidence": "low"}

    df = pd.DataFrame(sales)
    df = df[df["price"] > 0]

    if len(df) < 10:
        return {"elasticity": -1.2, "interpretation": "Estimated (insufficient data)", "confidence": "low"}

    log_price = np.log(df["price"].values)
    log_qty = np.log(np.maximum(df["quantity"].values, 0.1))

    # OLS log-log regression
    X = np.column_stack([np.ones(len(log_price)), log_price])
    try:
        coeffs = np.linalg.lstsq(X, log_qty, rcond=None)[0]
        elasticity = float(coeffs[1])
        r2 = float(1 - np.var(log_qty - X @ coeffs) / np.var(log_qty))
    except Exception:
        elasticity = -1.2
        r2 = 0

    if elasticity < -2:
        interpretation = "Highly elastic — demand very sensitive to price changes"
    elif elasticity < -1:
        interpretation = "Elastic — demand moderately sensitive to price"
    elif elasticity < 0:
        interpretation = "Inelastic — demand relatively insensitive to price"
    else:
        interpretation = "Giffen good — demand increases with price (unusual)"

    return {
        "elasticity": round(elasticity, 3),
        "interpretation": interpretation,
        "r2": round(r2, 3),
        "confidence": "high" if r2 > 0.5 else "medium" if r2 > 0.2 else "low",
        "data_points": len(df)
    }


def whatif_simulator(product_id: str, category: str, base_features: dict, scenarios: list) -> dict:
    """
    What-if simulator: test multiple price/stock/trend scenarios.
    scenarios: [{"name": "10% price drop", "price_change": -0.1, "trend_change": 0, "stock_change": 0}]
    """
    results = []
    base_price = base_features.get("price", 50)
    base_trend = base_features.get("trend_score", 50)
    base_stock = base_features.get("current_stock", 50)
    base_demand = base_features.get("avg_daily_sales_30d", 1) * 30

    for scenario in scenarios:
        new_price = base_price * (1 + scenario.get("price_change", 0))
        new_trend = base_trend + scenario.get("trend_change", 0)
        new_stock = base_stock * (1 + scenario.get("stock_change", 0))

        # Price elasticity effect
        price_ratio = new_price / max(base_price, 0.01)
        elasticity = -1.2  # default
        demand_multiplier = price_ratio ** elasticity

        # Trend effect
        trend_multiplier = 1 + (new_trend - base_trend) / 200

        new_demand = base_demand * demand_multiplier * trend_multiplier
        demand_change_pct = (new_demand - base_demand) / max(base_demand, 0.01) * 100

        # Revenue impact
        base_revenue = base_demand * base_price
        new_revenue = new_demand * new_price
        revenue_change_pct = (new_revenue - base_revenue) / max(base_revenue, 0.01) * 100

        results.append({
            "scenario_name": scenario.get("name", "Unnamed"),
            "price_change_pct": round(scenario.get("price_change", 0) * 100, 1),
            "new_price": round(new_price, 2),
            "predicted_demand": round(new_demand, 1),
            "demand_change_pct": round(demand_change_pct, 1),
            "new_revenue": round(new_revenue, 2),
            "revenue_change_pct": round(revenue_change_pct, 1),
            "recommendation": "Proceed" if revenue_change_pct > 0 else "Caution"
        })

    return {
        "base_demand": round(base_demand, 1),
        "base_price": base_price,
        "base_revenue": round(base_demand * base_price, 2),
        "scenarios": results
    }
